internal/provider/resource_tfe_opa_version.go

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package provider
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"strings"
+
+	tfe "github.com/hashicorp/go-tfe"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func resourceTFEOPAVersion() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceTFEOPAVersionCreate,
+		Read:   resourceTFEOPAVersionRead,
+		Update: resourceTFEOPAVersionUpdate,
+		Delete: resourceTFEOPAVersionDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: resourceTFEOPAVersionImporter,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"version": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"url": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"sha": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"official": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+			"enabled": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
+			"beta": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+			"deprecated": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+			"deprecated_reason": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  nil,
+			},
+		},
+	}
+}
+
+func resourceTFEOPAVersionCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(ConfiguredClient)
+
+	opts := tfe.AdminOPAVersionCreateOptions{
+		Version:          *tfe.String(d.Get("version").(string)),
+		URL:              *tfe.String(d.Get("url").(string)),
+		SHA:              *tfe.String(d.Get("sha").(string)),
+		Official:         tfe.Bool(d.Get("official").(bool)),
+		Enabled:          tfe.Bool(d.Get("enabled").(bool)),
+		Beta:             tfe.Bool(d.Get("beta").(bool)),
+		Deprecated:       tfe.Bool(d.Get("deprecated").(bool)),
+		DeprecatedReason: tfe.String(d.Get("deprecated_reason").(string)),
+	}
+
+	log.Printf("[DEBUG] Create new OPA version: %s", opts.Version)
+	v, err := config.Client.Admin.OPAVersions.Create(ctx, opts)
+	if err != nil {
+		return fmt.Errorf("Error creating the new OPA version %s: %w", opts.Version, err)
+	}
+
+	d.SetId(v.ID)
+
+	return resourceTFEOPAVersionUpdate(d, meta)
```

Comment:
By convention, our `Create` and `Update` methods generally return the `Read` method. Can you share more about why this method returns `Update`?

Reply:
Well... 😅 I was trying to follow the same approach as tfe_terraform_version provider as the policy version providers should be similar. Underneath the hood, they use the same tool_version api in atlas.

Reply:
I briefly looked into it, and I think `tfe_terraform_version` is in error here (probably just a typo) and all three resources should be returning a Read. Sorry for the misleading crib sheet!